File: testdata/fixtures/scala/CWE-918.scala
// CWE-918: SSRF via various Scala HTTP clients
import scalaj.http.Http
import dispatch._
import scala.io.Source
import play.api.libs.ws.WSClient
import javax.servlet.http.HttpServletRequest

object SsrfTest {
  def vulnDispatch(request: HttpServletRequest) = {
    val url = request.getParameter("url")
    // ruleid: scala-ssrf
    Http(url).asString
    // ruleid: scala-ssrf
    url(url)
    // ruleid: scala-ssrf
    Source.fromURL(url)
  }
  def vulnWs(request: HttpServletRequest, ws: WSClient) = {
    val target = request.getHeader("X-Target")
    // ruleid: scala-ssrf
    ws.url(target).get()
    val q = request.getQueryString()
    // ruleid: scala-ssrf
    Source.fromURI(q)
  }
  def safe(ws: WSClient) = {
    // ok: scala-ssrf
    Http("https://example.com").asString
    // ok: scala-ssrf
    url("https://example.com")
    // ok: scala-ssrf
    Source.fromURL("https://example.com")
    // ok: scala-ssrf
    ws.url("https://example.com").get()
  }
}

File: testdata/fixtures/rust/CWE-918.rs
// CWE-918: SSRF via reqwest with untrusted URL
use reqwest;

fn vuln() {
    let url = std::env::var("TARGET_URL").unwrap();
    // ruleid: rust-ssrf
    let _ = reqwest::get(url);
    let target = std::env::args().next().unwrap();
    // ruleid: rust-ssrf
    let _ = reqwest::blocking::get(target);
    let u = std::env::var("URL").unwrap();
    let propagated = u;
    // ruleid: rust-ssrf
    let _ = reqwest::get(propagated);
}

fn safe() {
    // ok: rust-ssrf
    let _ = reqwest::get("https://example.com");
    // ok: rust-ssrf
    let _ = reqwest::blocking::get("https://example.com");
    // ok: rust-ssrf
    let safe = "https://fixed.com";
    let _ = reqwest::get(safe);
}
